actors/builtin/miner/deadline_state.go

```diff
@@ -444,3 +446,145 @@ func (dl *Deadline) popExpiredPartitions(store adt.Store, until abi.ChainEpoch,
 
 	return popped, modified, nil
 }
+
+// RemovePartitions removes the specified partitions, shifting the remaining
+// ones to the left, and returning the live and dead sectors they contained.
+//
+// Returns an error if any of the partitions contained faulty sectors or early
+// terminations.
+func (dl *Deadline) RemovePartitions(store adt.Store, partitions *bitfield.BitField, quant QuantSpec) (
+	live, dead *abi.BitField, removedPower PowerPair, err error,
+) {
+	oldPartitions, err := dl.PartitionsArray(store)
+	if err != nil {
+		return nil, nil, NewPowerPairZero(), xerrors.Errorf("failed to load partitions: %w", err)
+	}
+
+	partitionCount := oldPartitions.Length()
+	toCompact, err := partitions.AllMap(partitionCount)
+	if err != nil {
+		// TODO: This may be an illegal argument error if we have too many partitions.
+		return nil, nil, NewPowerPairZero(), xerrors.Errorf("failed to expand partitions into map: %w", err)
+	}
+
+	newPartitions := adt.MakeEmptyArray(store)
+	allDeadSectors := make([]*bitfield.BitField, 0, len(toCompact))
+	allLiveSectors := make([]*bitfield.BitField, 0, len(toCompact))
+	removedPower = NewPowerPairZero()
+
+	// Define all of these out here to save allocations.
+	//
+	// TODO: This can load a bunch of partitions, and we won't flush
+	// in-between. This could use a lot of memory (100s of megabytes
+	// worst case?). This is probably fine, but something to be
+	// aware of.
+	var (
+		// TODO: this is lazy, but not that lazy. Ideally, we
+		// wouldn't do any decoding for untouched partitions.
+		lazyPartition cbg.Deferred
+		byteReader    bytes.Reader
+		partition     Partition
+	)
+	if err = oldPartitions.ForEach(&lazyPartition, func(partIdx int64) error {
+		// If we're keeping the partition as-is, append it to the new partitions array.
+		// TODO: consider using the lazy cbor unmarshaler to avoid unmarshaling partitions we're keeping.
+		if _, ok := toCompact[uint64(partIdx)]; !ok {
+			return newPartitions.AppendContinuous(&lazyPartition)
+		}
+
+		// Ok, actually unmarshal the partition.
+		byteReader.Reset(lazyPartition.Raw)
+		err := partition.UnmarshalCBOR(&byteReader)
+		byteReader.Reset(nil)
+		if err != nil {
+			return xerrors.Errorf("failed to decode partition %d: %w", partIdx, err)
+		}
+
+		// Don't allow compacting partitions with faulty sectors.
+		hasNoFaults, err := partition.Faults.IsEmpty()
```

Comment:
I'm rejecting partitions with faults (for now) because handling them is non-trivial. We need to:

1. Pull out any potential early terminations from the partitions termination queue. It'll just cost more gas.
2. Record faulty/recovering status and carry this over when we re-add the sectors. Unfortunately, this means changing all of our "add new sectors" logic (or adding parallel logic for re-adding known-faulty sectors).

In practice, I don't think this limitation will be problematic.

Reply:
https://github.com/filecoin-project/specs-actors/issues/756